In [1]:
from IPython.display import Markdown, display

with open("description.md", "r") as file:
    md_content = file.read()
display(Markdown(md_content))

# Problem 38

[**Pandigital Multiples**](https://projecteuler.net/problem=38)

## Description

Take the number 192 and multiply it by each of 1, 2, and 3:

$$
\begin{align}
192 \times 1 &= 192\\
192 \times 2 &= 384\\
192 \times 3 &= 576
\end{align}
$$

By concatenating each product we get the 1 to 9 pandigital, 192384576. We will call 192384576 the concatenated product of 192 and (1,2,3).

The same can be achieved by starting with 9 and multiplying by 1, 2, 3, 4, and 5, giving the pandigital, 918273645, which is the concatenated product of 9 and (1,2,3,4,5).

## Task

What is the largest 1 to 9 pandigital 9-digit number that can be formed as the concatenated product of an integer with $(1,2, \dots, n)$ where $n > 1$?


## Brute-force Solution

In [ ]:
import numpy as np


def main() -> int:
    max_result = 0

    for inverse_i in range(10, 2, -1):
        one_to_n_seq = np.arange(1, inverse_i)

        multiplier = 0
        result = 0

        while True:
            multiplier += 1
            array = one_to_n_seq * multiplier
            result = int("".join((str(i) for i in array)))

            if result >= 1e9:
                break

            if is_nine_pandigital(result) and result > max_result:
                max_result = result

                # print("new 9 pandigital found")
                # print(multiplier)
                # print(array)
                # print(result)

    return max_result


def is_nine_pandigital(num: int) -> bool:
    return len(str(num)) == 9 and set(str(num)) == set("123456789")

In [22]:
%%timeit
main()

24.3 ms ± 278 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [24]:
main()

932718654

## Optimized Solution

We already know about number 918273645 that is pandigital and result of concatenated product of 9 and (1,2,3,4,5).

We can brute-force all 4-digit numbers and check if they are pandigital when concatenated with their multiples by 1 and 2.

Multipliers below 9123 will not give pandigital numbers, similarly multipliers above 9876 will not give pandigital numbers.

In [ ]:
import numpy as np


def main2() -> int:
    multiplier = 9
    one_to_n_seq = np.arange(1, 6)
    array = one_to_n_seq * multiplier
    max_result = int("".join((str(i) for i in array)))

    one_to_n_seq = np.arange(1, 3)
    multiplier = 8
    result = 0

    for multiplier in range(9123, 9876):
        array = one_to_n_seq * multiplier
        result = int("".join((str(i) for i in array)))

        if result >= 1e9:
            break

        if is_nine_pandigital(result) and result > max_result:
            max_result = result

    return max_result


def is_nine_pandigital(num: int) -> bool:
    return len(str(num)) == 9 and set(str(num)) == set("123456789")

In [53]:
%%timeit
main2()

2 ms ± 56.6 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [49]:
main2()

932718654